src/flow_solver/flow_solver.cpp

```diff
@@ -426,40 +472,58 @@ int FlowSolver<dim,nstate>::run() const
         //----------------------------------------------------
         // Time advancement loop with on-the-fly post-processing
         //----------------------------------------------------
+        double next_time_step = time_step;
         pcout << "Advancing solution in time... " << std::endl;
         pcout << "Timer starting. " << std::endl;
         dealii::Timer timer(this->mpi_communicator,false);
         timer.start();
-        while((ode_solver->current_time) < (final_time - 1E-13)) //comparing to 1E-13 to avoid taking an extra timestep
+        while(ode_solver->current_time < final_time)
         {
-            // update adaptive time step
-            if(flow_solver_param.adaptive_time_step == true) {
-                time_step = flow_solver_case->get_adaptive_time_step(dg);
-                flow_solver_case->set_time_step(time_step);
+            time_step = next_time_step; // update time step
+
+            // check if we need to decrease the time step
+            if((ode_solver->current_time+time_step) > final_time) {
+                // decrease time step to finish exactly at specified final time
+                time_step = final_time - ode_solver->current_time;
+            } else if (this->output_solution_at_exact_fixed_times && (this->do_output_solution_at_fixed_times && (this->number_of_fixed_times_to_output_solution > 0))) { // change this to some parameter
+                const double next_time = ode_solver->current_time + time_step;
+                const double desired_time = this->output_solution_fixed_times[index_of_current_desired_fixed_time_to_output_solution];
+                // // Check if current time is an output time
```

Comment:
Not sure if this line will appear in the doxygen doc. Maybe remove the extra `//`.

Reply:
fixed